tests/integration_tests/platforms.py

```diff
@@ -0,0 +1,232 @@
+from tests.integration_tests import integration_settings
+
+try:
+    from typing import Callable, Optional
+except ImportError:
+    pass
+
+current_image = None
+
+
+def set_current_image(image):
+    global current_image
+    current_image = image
+
+
+class IntegrationClient(ABC):
+    client = None  # type: Optional[BaseCloud]
+    instance = None  # type: Optional[BaseInstance]
+    datasource = None  # type: Optional[str]
+    use_sudo = True
+
+    def __init__(self, user_data=None, instance_type=None, wait=True,
+                 settings=integration_settings, launch_kwargs=None):
+        self.user_data = user_data
+        self.instance_type = settings.INSTANCE_TYPE if \
+            instance_type is None else instance_type
+        self.wait = wait
+        self.settings = settings
+        self.launch_kwargs = launch_kwargs if launch_kwargs else {}
+        self.client = self._get_client()
+
+    @abstractmethod
+    def _get_client(self):
+        raise NotImplementedError
+
+    def _get_image(self):
+        if current_image:
+            return current_image
+        image_id = self.settings.OS_IMAGE
+        try:
+            image_id = self.client.released_image(self.settings.OS_IMAGE)
+        except (ValueError, IndexError):
+            pass
+        return image_id
+
+    def launch(self):
+        if self.settings.EXISTING_INSTANCE_ID:
+            print('Not launching instance due to EXISTING_INSTANCE_ID. '
+                  'Instance id: {}'.format(self.settings.EXISTING_INSTANCE_ID))
+            self.instance = self.client.get_instance(
+                self.settings.EXISTING_INSTANCE_ID
+            )
+            return
+        image_id = self._get_image()
+        launch_args = {
+            'image_id': image_id,
+            'user_data': self.user_data,
+            'wait': self.wait,
+        }
+        if self.instance_type:
+            launch_args['instance_type'] = self.instance_type
+        launch_args.update(self.launch_kwargs)
+        self.instance = self.client.launch(**launch_args)
+        print('Launched instance: {}'.format(self.instance))
+
+    def destroy(self):
+        self.instance.delete()
+
+    def exec(self, command):
+        return self.instance.execute(command)
+
+    def pull_file(self, remote_file, local_file):
+        self.instance.pull_file(remote_file, local_file)
+
+    def push_file(self, local_path, remote_path):
+        self.instance.push_file(local_path, remote_path)
+
+    def pull_from_file(self, remote_path) -> str:
+        tmp_file = NamedTemporaryFile('r')
+        self.pull_file(remote_path, tmp_file.name)
+        with tmp_file as f:
+            contents = f.read()
+        return contents
+
+    def push_to_file(self, remote_path, contents: str):
+        with NamedTemporaryFile('w', delete=False) as tmp_file:
+            tmp_file.write(contents)
+        self.push_file(tmp_file.name, remote_path)
+        os.unlink(tmp_file.name)
```

Comment:
> Could there be problems with pushing a still open file? We'd have to at least flush first, correct?

Hmm, great point; probably and probably.

In that case, perhaps:

```py
with contextlib.ExitStack() as stack:
    stack.callback(os.unlink, tmp_file.name)
    self.push_file(...)
```

would do the trick?

(Also, thinking about it, with `delete=False`, using `NamedTemporaryFile` as a context manager doesn't do anything, so probably just `tmp_file = NamedTemporaryFile('w', delete=False)` would make it clearer that we aren't using the built-in cleanup.)

> This implementation is a little lazy, but I wasn't too concerned with occasionally leaving a file in /tmp/ in the case where the push fails.

The failure case I'm concerned about is on a Jenkins slave, if we don't notice failures for hours or days then we could end up with an awful lot of these files in /tmp (and if they're cloud-init debs then they're ~0.5MB, so they can add up pretty fast). Not a huge deal, agreed.

Reply:
How about this?
```python
    def write_to_file(self, remote_path, contents: str):
        with NamedTemporaryFile('w', delete=False) as tmp_file:
            tmp_file.write(contents)

        try:
            self.push_file(tmp_file.name, remote_path)
        finally:
            os.unlink(tmp_file.name)
```

I still want the `NamedTemporaryFile` context manager so the file gets closed automatically (otherwise I need a try...finally around the write). Your `ExitStack()` code seems like a convenient way to get a generic context manager, but it feels a bit golf-y to me and it wasn't immediately clear to me what was happening, so I think I prefer a try...finally.

I can't combine them into a single `ExitStack()` block because I need the file object to add a callback, but since it's a stack, the unlink would happen before the close.

Reply:
LMAO I had forgotten about the existence of try/finally apparently. Yeah, that's way better.